src/BattleServer/battle.cpp

```diff
@@ -3238,235 +3289,385 @@ int BattleSituation::calculateDamage(int p, int t)
+        /* Calculate the multiplier for two turn attacks */
+        if (pokeMemory(t).contains("VulnerableMoves") && pokeMemory(t).value("Invulnerable").toBool()) {
+            QList<int> vuln_moves = pokeMemory(t)["VulnerableMoves"].value<QList<int> >();
+            QList<int> vuln_mults = pokeMemory(t)["VulnerableMults"].value<QList<int> >();
+            for (int i = 0; i < vuln_moves.size(); i++) {
+                if (vuln_moves[i] == attackused) {
+                    power *= vuln_mults[i];
+                }
+            }
+        }
 
-    /* Used by Pledges to use a special attack, the sum of both */
-    if (move.contains("AttackStat")) {
-        attack = move.value("AttackStat").toInt();
-        move.remove("AttackStat");
-    }
-
-    attack = std::min(attack, 65535);
-
-    if ( (attackused == Move::Explosion || attackused == Move::Selfdestruct) && gen() <= 4) {
-        /* explosion / selfdestruct */
-        def/=2;
-        if (def == 0)
-            // prevent division by zero
-            def = 1;
-    }
-
-    int stab = turnMem(p).stab;
-    int typemod = turnMem(p).typeMod;
-    int randnum = randint(16) + 85;
-    //Spit Up
-    if (attackused == Move::SpitUp && gen().num == 3) {
-        randnum = 100;
-    }
-
-    int ch;
-    if (gen() <= 5) {
-        ch = 4+(crit * (1+hasWorkingAbility(p,Ability::Sniper)))*4;
-    } else {
-        //1.5, 2.25 with sniper
-        ch = 4+crit*(2+3*(hasWorkingAbility(p,Ability::Sniper)));
-    }
-
-    /*** WARNING ***/
-    /* The peculiar order here is caused by the fact that helping hand applies before item boosts,
-      but item boosts are decided (not applied) before acrobat, and acrobat needs to modify
-      move power (not just power variable) because of technician which relies on it */
+        /* Helping Hand, Gen 4 is Power. Gen 3 is Damage */
+        if (gen() > 3 && move.contains("HelpingHanded")) {
+            power = power * 3 / 2;
+        }
 
-    calleffects(p,t,"BasePowerModifier");
-    callieffects(p,t,"BasePowerModifier");
-    if (turnMemory(p).value("GemActivated").toBool()) {
-        gen() < 6 ? chainBp(p, 2048) : chainBp(p, 1229);
-    }
+        /* Move *///Moves: Facade, Brine
+        calleffects(p,t,"BasePowerModifier");
+        power = floorMod(power);
 
-    /* The Acrobat thing is here because it's supposed to activate after gem Consumption */
-    if (attackused == Move::Acrobatics && poke.item() == Item::NoItem) {
-        tmove(p).power *= 2;
-    }
+        /* Item *///Items: Muscle Band, Wise Glasses, Type boosting items, Adamant/Lustrous/Griseous Orb
+        callieffects(p,t,"BasePowerModifier");
+        power = floorMod(power);
 
-    callaeffects(p,t,"BasePowerModifier");
-    callaeffects(t,p,"BasePowerFoeModifier");
+        //TODO: Make this work
+        /* Charge */// Can't be called via ChainBP else it will be out of order
+        if (move.contains("Charged")) {
+            power *= 2;
```

Comment:
```
            calleffects(player, target, "BeforeCalculatingDamage");
            /* For charge */
            callpeffects(player, target, "BeforeCalculatingDamage");
``` 

That's the current code. Since the functions are stored in `poke(b,s)` then `callpeffects` needs to be used. Current only `calleffects` is used regarding base power modifiers. Does charge even work in gen 5?